### PR TITLE
Add Audio Queries and Commands

### PR DESCRIPTION
**Pull request by ZokirS**

### CODE REVIEW

File: Elevel/Application/Features/AudioFeatures/Commands/CreateAudioCommand.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+{
+    public class CreateAudioCommand: IRequest<Guid>
+    {
+        public string AudioFilePath { get; set; }
+        public Level Level { get; set; }
+        public DateTimeOffset CreationDate { get; set; }
+        public ICollection<Question> Questions { get; set; }
+        public ICollection<Test> Tests { get; set; }
+        public class CreateAudioCommandHandler: IRequestHandler<CreateAudioCommand, Guid>
+        {
```
Comment (andreyrogov8) [inline Elevel/Application/Features/AudioFeatures/Commands/CreateAudioCommand.cs:20]: We should move properties to a separate file

File: Elevel/Application/Features/AudioFeatures/Commands/UpdateAudioCommand.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    public class UpdateAudioCommand: IRequest<Guid>
+    {
+        public Guid Id { get; set; }
+        public string AudioFilePath { get; set; }
+        public Level Level { get; set; }
+        public DateTimeOffset CreationDate { get; set; }
+        public ICollection<Question> Questions { get; set; }
+        public ICollection<Test> Tests { get; set; }
+        public class UpdateAudioCommandHandler: IRequestHandler<UpdateAudioCommand, Guid>
+        {
```
Comment (andreyrogov8) [inline Elevel/Application/Features/AudioFeatures/Commands/UpdateAudioCommand.cs:22]: You're duplicating code 
The same properties are contained in CreateAudioCommand.cs

Comment (parys) [reply Elevel/Application/Features/AudioFeatures/Commands/UpdateAudioCommand.cs:22]: no, you should not operate with domain model for input-output outside business logic layer (Applicaition)

**Review by 4500zenja1 — DISMISSED** [review]

**Comment by 4500zenja1** [comment]
Why you need to create the Auditions table, if it was ALREADY in IApplicationDbContext?

**Review by 4500zenja1 — APPROVED** [review]

File: Elevel/Application/Features/AudioFeatures/DeleteAudiotionCommand.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+
+namespace Elevel.Application.Features.AudioFeatures
+{
+   public class DeleteAudiotionCommand
+    {
+        public class Request : IRequest<Response>
+        {
+            public Guid Id{ get; set; }
+            public string AudioFilePath { get; set; }
```
Comment (parys) [inline Elevel/Application/Features/AudioFeatures/DeleteAudiotionCommand.cs:17]: for delete you just need Id

File: Elevel/Application/Features/AudioFeatures/DeleteAudiotionCommand.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            public DateTimeOffset CreationDate { get; set; }
+        }
+
+        public class Handler : IRequestHandler<Request, Response>
+        {
+            private readonly IApplicationDbContext _context;
+            private readonly IMapper _mapper;
```
Comment (parys) [inline Elevel/Application/Features/AudioFeatures/DeleteAudiotionCommand.cs:24]: you don't need Mapper for delete command, you are not using it

File: Elevel/Application/Features/AudioFeatures/CreateAudiotionCommand.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+using System.Threading.Tasks;
+using MediatR;
+using Elevel.Application.Interfaces;
+using AutoMapper;
+using System.Threading;
+using Elevel.Domain.Models;
+
+namespace Elevel.Application.Features.AudioFeatures
+{
+    public class CreateAudiotionCommand
```
Comment (parys) [inline Elevel/Application/Features/AudioFeatures/CreateAudiotionCommand.cs:11]: mistype, audiOtion

**Review by 4500zenja1 — APPROVED** [review]

**Review by ristomin — APPROVED** [review]

File: .gitignore [hunk 6]
```diff
@@ -276,4 +276,5 @@ ServiceFabricBackup/
 *.rptproj.rsuser
 *- [Bb]ackup.rdl
 *- [Bb]ackup ([0-9]).rdl
-*- [Bb]ackup ([0-9][0-9]).rdl
\ No newline at end of file
+*- [Bb]ackup ([0-9][0-9]).rdl
+/Elevel/Application/DependencyInjection1.cs
```
Comment (parys) [inline .gitignore:280]: why?

File: Elevel/Application/Features/AudioFeatures/GetAuditionByIdQuery.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+                _mapper = mapper;
+            }
+            public async Task<Response> Handle(Request request, CancellationToken cancellationToken)
+            {
+                var audition = await _context.Auditions.AsNoTracking()
+                    .ProjectTo<Response>(_mapper.ConfigurationProvider)
+                    .FirstOrDefaultAsync(m => m.Id == request.Id, cancellationToken);
+                if(audition is null)
+                {
+                    throw new NullReferenceException();
```
Comment (parys) [inline Elevel/Application/Features/AudioFeatures/GetAuditionByIdQuery.cs:40]: please not throw NRE yourself

File: Elevel/Application/Features/AudioFeatures/GetAuditionListQuery.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            public async Task<Response> Handle(Request request, CancellationToken cancellationToken)
+            {
+                var auditionList = _context.Auditions.AsNoTracking()
+                    .Include(c => c.Id)
+                    .Include(c => c.AudioFilePath)
+                    .Include(c => c.CreationDate)
+                    .OrderBy(c => c.CreationDate);
+
+                var audition = new List<AuditionDto>();
+                foreach (var item in auditionList)
```
Comment (parys) [inline Elevel/Application/Features/AudioFeatures/GetAuditionListQuery.cs:40]: please look how to use PagedResult

File: Elevel/Persistence/Migrations/20210713202807_init.Designer.cs [hunk 9]
```diff
@@ -1,613 +0,0 @@
-﻿// <auto-generated />
```
Comment (parys) [inline Elevel/Persistence/Migrations/20210713202807_init.Designer.cs:1]: why delete?

**Review by parys — CHANGES_REQUESTED** [review]
please fixed comments

File: Elevel/Application/Features/AudioFeatures/GetAuditionByIdQuery.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+using System.Threading.Tasks;
+
+namespace Elevel.Application.Features.AudioFeatures
+{
+    public class GetAuditionByIdQuery
+    {
+        public class Request: IRequest<Response>
+        {
+            public Guid Id { get; set; }
+            public string AudioFilePath { get; set; }
```
Comment (DShcharbakou) [inline Elevel/Application/Features/AudioFeatures/GetAuditionByIdQuery.cs:20]: Why do you need AudioFilePath and CreationDate in Get request?

File: Elevel/Application/Features/AudioFeatures/GetAuditionListQuery.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+using Microsoft.EntityFrameworkCore;
+using Elevel.Application.Pagination;
+
+namespace Elevel.Application.Features.AudioFeatures
+{
+    public class GetAuditionListQuery
+    {
+        public class Request : PagedQueryBase, IRequest<Response>
+        {
+            public Guid Id { get; set; }
```
Comment (DShcharbakou) [inline Elevel/Application/Features/AudioFeatures/GetAuditionListQuery.cs:19]: For all auditions do not need **Id** and **AudioFilePath**

File: Elevel/Application/Features/AudioFeatures/GetAuditionListQuery.cs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+            public Handler(IApplicationDbContext context, IMapper mapper)
+            {
+                _context = context;
+                _mapper = mapper;
+            }
+            public async Task<Response> Handle(Request request, CancellationToken cancellationToken)
+            {
+                var auditionList = _context.Auditions.AsNoTracking()
+                    .Include(c => c.Id)
+                    .Include(c => c.AudioFilePath)
```
Comment (DShcharbakou) [inline Elevel/Application/Features/AudioFeatures/GetAuditionListQuery.cs:35]: https://www.entityframeworktutorial.net/efcore/querying-in-ef-core.aspx read this article

File: Elevel/Elevel.Api/appsettings.json [hunk 13]
```diff
@@ -9,7 +9,7 @@
 
   "AllowedHosts": "*",
   "ConnectionStrings": {
-    "DefaultConnection": "Server=localhost\\SQLEXPRESS03;Database=Elevel;Trusted_Connection=True;MultipleActiveResultSets=true",
+    "DefaultConnection": "Server=localhost\\SQLEXPRESS;Database=Elevel;Trusted_Connection=True;MultipleActiveResultSets=true",
```
Comment (DShcharbakou) [inline Elevel/Elevel.Api/appsettings.json:12]: Why did you change Connection string?

**Review by DShcharbakou — CHANGES_REQUESTED** [review]

File: Elevel/Application/Features/AudioFeatures/CreateAuditionCommand.cs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+using Elevel.Domain.Models;
+
+namespace Elevel.Application.Features.AudioFeatures
+{
+    public class CreateAuditionCommand
+    {
+        public class Request : IRequest<Response>
+        {
+            public string AudioFilePath { get; set; }
+            public DateTimeOffset CreationDate { get; set; }
```
Comment (parys) [inline Elevel/Application/Features/AudioFeatures/CreateAuditionCommand.cs:16]: Creation date should be set at your handler, clint should not send it.
and what path will be at audioFilePath?

File: Elevel/Application/Features/AudioFeatures/UpdateAuditionCommand.cs [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace Elevel.Application.Features.AudioFeatures
+{
+    public class UpdateAuditionCommand
+    {
+        public class Request: IRequest<Response>
+        {
+            public Guid Id { get; set; }
+            public string AudioFilePath { get; set; }
+            public DateTimeOffset CreationDate { get; set; }
```
Comment (parys) [inline Elevel/Application/Features/AudioFeatures/UpdateAuditionCommand.cs:17]: you can't change creationDate on Update

File: Elevel/Elevel.Api/Controllers/AuditionController.cs [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+namespace Elevel.Api.Controllers
+{
+    [Route("[controller]/[action]")]
+    public class AuditionController : BaseApiController
+    {
+        private readonly IMediator _mediator;
+        public AuditionController(IMediator mediator)
+        {
+            _mediator = mediator;
```
Comment (parys) [inline Elevel/Elevel.Api/Controllers/AuditionController.cs:18]: you already have Mediator for you in BaseApiController class, please use it
+ Add Authorize attribute to you controller

File: Elevel/Application/Features/AudioFeatures/CreateAuditionCommand.cs [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+            private readonly IApplicationDbContext _context;
+            private readonly IMapper _mapper;
+            public Handler(IApplicationDbContext context, IMapper mapper)
+            {
+                _context = context;
+                _mapper = mapper;
+            }
+            public async Task<Response> Handle(Request request, CancellationToken cancelationtoken)
+            {
+                request.CreationDate = DateTime.Now;
```
Comment (parys) [inline Elevel/Application/Features/AudioFeatures/CreateAuditionCommand.cs:29]: please use DateTimeOffset.NowUtc
and if you set this property here why we have it in request?

File: Elevel/Application/Features/AudioFeatures/UpdateAuditionCommand.cs [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace Elevel.Application.Features.AudioFeatures
+{
+    public class UpdateAuditionCommand
+    {
+        public class Request: IRequest<Response>
+        {
+            public Guid Id { get; set; }
+            public string AudioFilePath { get; set; }
+            public DateTimeOffset UpdatedDate { get; set; }
```
Comment (parys) [inline Elevel/Application/Features/AudioFeatures/UpdateAuditionCommand.cs:17]: do we have update date in model? it can be setup in handler during update too

File: Elevel/Elevel.Api/Controllers/BaseApiController.cs [hunk 19]
```diff
@@ -13,7 +13,7 @@ namespace Elevel.Api.Controllers
     [ApiController]
     public abstract class BaseApiController : ControllerBase
     {
-        private IMediator _mediator;
+        internal IMediator _mediator;
```
Comment (parys) [inline Elevel/Elevel.Api/Controllers/BaseApiController.cs:16]: you should not change this
you can use Mediator on row below

Comment (ZokirS) [reply Elevel/Elevel.Api/Controllers/BaseApiController.cs:16]: Do you mean i should add extra IMediator field with internal modificator?



> you should not change this
> you can use Mediator on row below

File: Elevel/Application/Features/AudioFeatures/GetAuditionListQuery.cs [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+            private readonly IMapper _mapper;
+            public Handler(IApplicationDbContext context, IMapper mapper)
+            {
+                _context = context;
+                _mapper = mapper;
+            }
+            public async Task<Response> Handle(Request request, CancellationToken cancellationToken)
+            {
+                var audition = _context.Auditions.AsNoTracking()
+                    .Include(x => x.Id);
```
Comment (DShcharbakou) [inline Elevel/Application/Features/AudioFeatures/GetAuditionListQuery.cs:31]: https://www.entityframeworktutorial.net/efcore/querying-in-ef-core.aspx

**Review by DShcharbakou — CHANGES_REQUESTED** [review]
